### PR TITLE
Merge feature/use-dllimport-generator to main

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Merge the DllImportGenerator experiment to the main branch for .NET 7.

This should be merged with "Create a merge commit"

ASP.NET JSON macro-benchmark: https://github.com/dotnet/runtime/pull/59579#issuecomment-938878840
Libraries micro-benchmarks: https://github.com/dotnet/runtime/pull/59579#issuecomment-940286862

Items we will address once the initial merge happens: https://github.com/dotnet/runtime/issues/60212

cc @stephentoub

### CODE REVIEW

**Comment by am11** [comment]
Do we want to keep the new `src/samples/` directory, or move it to https://github.com/dotnet/samples?

**Comment by jkoritzinsky** [comment]
The new src/samples directory is partially meant for quick and simple testing with native debuggers without having to go through `dotnet test`. Very similar to AndroidSampleApp and the other samples in src/mono/samples/.

**Comment by jkotas** [comment]
> The new src/samples directory is partially meant for quick and simple testing with native debuggers 

Should it have more appropriate name (e.g. `TestApps`) to capture the intent and make it obvious that these are not meant to be user-facing samples?

Also, should it have less stuff in it (e.g. just a single signature)?

File: src/libraries/System.Runtime.InteropServices/tests/TestAssets/NativeExports/NativeExports.csproj [hunk 1]
```diff
@@ -17,4 +20,97 @@
     <ProjectReference Include="..\SharedTypes\SharedTypes.csproj" />
   </ItemGroup>
 
+  <Target Name="GetLinuxCrossBuildArgumentsForDNNE"
```
Comment (am11) [inline src/libraries/System.Runtime.InteropServices/tests/TestAssets/NativeExports/NativeExports.csproj:23]: Can the DNNE build moved under `src/tests`where common native infra is already wired up? We can add a simple cmake file for DNNE project there (as we have for other external projects  libunwind, brotli etc.), without replicating the script code so implicitly.

Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/tests/TestAssets/NativeExports/NativeExports.csproj:23]: We'd rather not move the DNNE project stuff over to the `src/tests` tree as that will prevent us from using XUnit for our tests easily. We're also trying to keep the DNNE native component very slim and as transparent to the build as possible, and moving it into the CMake build for tests would make it not transparent due to build dependencies (the native C file is generated from the compiled managed assembly)

**Comment by jkoritzinsky** [comment]
I've run the ASP.NET JSON macrobenchmark on this branch and compared it to main, and we see a minor performance improvement in steady state performance with this branch.

<summary>Baseline</summary>
<details>

| application           |                                |
| --------------------- | ------------------------------ |        
| CPU Usage (%)         | 97                             |        
| Cores usage (%)       | 1,166                          |        
| Working Set (MB)      | 182                            |        
| Private Memory (MB)   | 771                            |        
| Build Time (ms)       | 13,179                         |
| Start Time (ms)       | 184                            |        
| Published Size (KB)   | 109,864                        |        
| .NET Core SDK Version | 5.0.401                        |        
| ASP.NET Core Version  | 5.0.10+b92c9f5                 |        
| .NET Runtime Version  | 7.0.0-alpha.1.21507.12+8fdd22d |        


| load                   |                |
| ---------------------- | -------------- |
| CPU Usage (%)          | 62             |
| Cores usage (%)        | 745            |
| Working Set (MB)       | 50             |
| Private Memory (MB)    | 358            |
| Build Time (ms)        | 7,140          |
| Start Time (ms)        | 0              |
| Published Size (KB)    | 76,412         |
| .NET Core SDK Version  | 3.1.413        |
| ASP.NET Core Version   | 3.1.19+b908e91 |
| .NET Runtime Version   | 3.1.19+aae0024 |
| First Request (ms)     | 123            |
| Requests/sec           | 469,558        |
| Requests               | 7,090,260      |
| Mean latency (ms)      | 0.79           |
| Max latency (ms)       | 48.15          |
| Bad responses          | 0              |
| Socket errors          | 0              |
| Read throughput (MB/s) | 68.07          |
| Latency 50th (ms)      | 0.50           |
| Latency 75th (ms)      | 0.58           |
| Latency 90th (ms)      | 0.72           |
| Latency 99th (ms)      | 10.98          |

</details>

<summary>This PR</summary>
<details>

| application           |                                |
| --------------------- | ------------------------------ |        
| CPU Usage (%)         | 97                             |        
| Cores usage (%)       | 1,168                          |        
| Working Set (MB)      | 185                            |        
| Private Memory (MB)   | 733                            |        
| Build Time (ms)       | 10,301                         |        
| Start Time (ms)       | 197                            |
| Published Size (KB)   | 110,366                        |        
| .NET Core SDK Version | 5.0.401                        |        
| ASP.NET Core Version  | 5.0.10+b92c9f5                 |        
| .NET Runtime Version  | 7.0.0-alpha.1.21507.14+78f4fa8 |        


| load                   |           |
| ---------------------- | --------- |
| CPU Usage (%)          | 63        |
| Cores usage (%)        | 754       |
| Working Set (MB)       | 38        |
| Private Memory (MB)    | 358       |
| Start Time (ms)        | 0         |
| First Request (ms)     | 129       |
| Requests/sec           | 471,137   |
| Requests               | 7,114,106 |
| Mean latency (ms)      | 0.78      |
| Max latency (ms)       | 52.35     |
| Bad responses          | 0         |
| Socket errors          | 0         |
| Read throughput (MB/s) | 68.30     |
| Latency 50th (ms)      | 0.50      |
| Latency 75th (ms)      | 0.58      |
| Latency 90th (ms)      | 0.71      |
| Latency 99th (ms)      | 10.73     |
</details>

**Comment by AaronRobinsonMSFT** [comment]
> and there is no significant gain in performance either.

@kasperk81 This is a short term and intentional non-goal. The reasoning is we want to ensure a high level of confidence and compatibility for v1. To that end we did do some optimizations, but very few. As we gain confidence with how this methodology works we are free to try optimizations that are targeted as opposed to the "everyone gets it" design with the built-in system. As we get further into .NET 7 and onto .NET 8 we are considering opportunities for better performance in this area.

**Review by elinor-fung — APPROVED** [review]

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Review by AaronRobinsonMSFT — APPROVED** [review]
LGTM

**Pull request merged**